lib/DB/Schema/ResultSet/Course.pm

```diff
@@ -262,36 +335,202 @@ This gets the Course Settings for a course
+		unless defined($global_setting);
+
+	my $course  = $self->getCourse(info => getCourseInfo($args{info}), as_result_set => 1);
+	my $setting = $course->course_settings->find({ setting_id => $global_setting->setting_id });
+
+	return $setting if $args{as_result_set};
+	my $setting_to_return =
+		$args{merged}
+		? { $setting->get_inflated_columns, $setting->global_setting->get_inflated_columns }
+		: { $setting->get_inflated_columns };
 
-	my $course_settings  = getDefaultCourseValues();
-	my $settings_from_db = { $course->course_settings->get_inflated_columns };
-	return mergeCourseSettings($course_settings, $settings_from_db);
+	# value and default_value are decoded from JSON as a hash.  Return to a value.
+	for my $key (qw/default_value value/) {
+		$setting_to_return->{$key} = $setting_to_return->{$key}->{value} if defined($setting_to_return->{$key});
+	}
+	return $setting_to_return;
 }
 
-sub updateCourseSettings ($self, %args) {
-	my $course = $self->getCourse(info => $args{info}, as_result_set => 1);
-	validateCourseSettings($args{settings});
+=pod
+
+=head2 updateCourseSetting
+
+Update a single course setting.
+
+=head3 input
+
+=over
+
+=item * C<info> which is a hash containing information about the course (either a
+C<course_id> or C<course_name>) and a setting (either a C<setting_id> or C<setting_name>).
+
+=item * C<params> the updated value of the course setting.
+
+=item * C<merged>, a boolean on whether the course setting is merged with its corresponding
+global setting.
+
+=item * C<$as_result_set>, a boolean if the return is to be a result_set
+
+=back
+
+=head3 output
+
+A single course setting as either a hashref or a C<DBIx::Class::ResultSet::CourseSetting> object.
+
+=cut
+
+sub updateCourseSetting ($self, %args) {
+	my $course = $self->getCourse(info => getCourseInfo($args{info}), as_result_set => 1);
+
+	my $global_setting = $self->getGlobalSetting(info => getSettingInfo($args{info}));
+
+	my $course_setting = $course->course_settings->find({
+		setting_id => $global_setting->{setting_id}
+	});
 
-	my $current_settings = { $course->course_settings->get_inflated_columns };
-	my $updated_settings = mergeCourseSettings($current_settings, $args{settings});
+	# Check that the setting is valid.
 
-	my $cs = $course->course_settings->update($updated_settings);
-	return mergeCourseSettings(getDefaultCourseValues(), { $cs->get_inflated_columns });
+	my $params = {
+		course_id  => $course->course_id,
+		setting_id => $global_setting->{setting_id},
+		value      => { value => $args{params}{value} }
+	};
+
+	# remove the following fields before checking for valid settings:
+	delete $global_setting->{$_} for (qw/setting_id course_id/);
+
+	isValidSetting($global_setting, $params->{value}{value});
+
+	# The course_id must be deleted to ensure it is written to the database correctly.
+	delete $params->{course_id} if defined($params->{course_id});
```

Comment:
This is awfully confusing. Our setting validation subroutine requires some keys to be deleted, and others to be added, which we don't actually want to use? Maybe we should re-think how `isValidSetting` is implemented, because this looks bizarre (we're defining `$params`, apparently with key-values that we don't even want -- just so we can validate?)

Reply:
this is cleaned up.  I'm ignoring any invalid fields now (must like we are doing throughout the codebase).